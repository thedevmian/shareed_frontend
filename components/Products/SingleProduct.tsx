import { useGetProductQuery } from "@/graphql/types";
import Head from "next/head";
import styled from "styled-components";
import formatMoney from "../../lib/products/formatMoney";
import AddToBagButton from "./AddToBagButton";
import AddToFavoriteButton from "./AddToFavoriteButton";
import ZoomMainPicture from "./ZoomMainPicture";
import ZoomPicture from "./ZoomPicture";
import { ISingleProductProps } from "pages/product/[id]";

const ProductContainer = styled.main`
  width: 100%;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  @media (min-width: 768px) {
    background-color: white;
  }
`;

const Heading = styled.h1`
  margin: 0.5rem 0;
  font-size: 2rem;
  font-weight: 400;
`;

const Description = styled.p`
  font-size: 0.9rem;
  width: 100%;
  line-height: 2rem;
  margin-bottom: 3rem;
  color: var(--main-text-color);

  @media (min-width: 1440px) {
    width: 50%;
  }
`;

const Price = styled.p`
  font-size: 1rem;
  font-weight: bold;
  color: var(--main-text-color);
`;

const Span = styled.span`
  font-size: 0.8rem;
  color: var(--main-text-color);
`;
const InfoContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
  padding: 2rem;
  width: 100%;

  @media (min-width: 768px) {
    margin-left: 2rem;
    width: 60%;
  }

  @media (min-width: 1024px) {
    margin-left: 4rem;
    width: 40%;
  }
`;

const ImageWrapper = styled.div`
  @media (min-width: 768px) {
    width: 50%;
    height: auto;
  }
`;

const LineWrapper = styled.div`
  width: 100%;
  margin: 0;
  position: relative;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  gap: 0.5rem;
`;

const MainSection = styled.section`
  display: flex;
  flex-direction: column;

  @media screen and (min-width: 1024px) {
    width: 80%;
    flex-direction: row;
  }
`;

const AnotherPhotoContainer = styled.div`
  width: 80%;
  margin-top: 4rem;
  margin-bottom: 4rem;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: flex-start;
  gap: 4rem;
`;

const SingleProduct = ({ id }: ISingleProductProps) => {
  const { loading, error, data } = useGetProductQuery({
    variables: { id },
  });
  if (loading) return <p>Loading...</p>;
  if (error) return <p data-testid="error">Error :{error.message}</p>;
  if (data) {
    const { name, photo, description, id, price } = data.product!;
    if (!photo) {
      return <p>No photo</p>;
    }
    const restofPhotos = photo.slice(1, photo.length);
    return (
      <ProductContainer data-testid="single-product">
        <Head>
          <title>Shareed - {name}</title>
        </Head>
        <MainSection>
          <ImageWrapper>
            <ZoomMainPicture
              src={photo[0].image?.publicUrlTransformed as string}
              alt={photo[0].image?.filename as string}
            />
          </ImageWrapper>
          <InfoContainer>
            <LineWrapper>
              <Heading>{name}</Heading>
              <AddToFavoriteButton />
            </LineWrapper>
            <Description>{description}</Description>
            <LineWrapper>
              <Price>{formatMoney(price as number)}</Price>
              <Span>(VAT included)</Span>
            </LineWrapper>
            <Description>
              Mid-washed light indigo denim jeans.
              <br />
              Straight fit Classic waist
              <br />
              Straight leg 18 cm bottom leg width (Size 30) 5 pockets Tobacco
              topstitches
              <br />
              Buttoned fly
              <br />
              Metal rivets and buttons in antic black color ami black leather
              patch on the back
              <br />
              Eco-friendly washing reducing its environmental impact
              <br />
              Main fabric:100% Cotton
            </Description>
            <AddToBagButton productId={id} />
          </InfoContainer>
        </MainSection>
        <AnotherPhotoContainer>
          {restofPhotos.map((photo) => (
            <ZoomPicture
              key={photo.id}
              src={photo.image?.publicUrlTransformed as string}
              alt={photo.image?.filename as string}
            />
          ))}
        </AnotherPhotoContainer>

        <div>
          <Heading>You might also like</Heading>
        </div>
      </ProductContainer>
    );
  } else {
    return <></>;
  }
};

export default SingleProduct;
