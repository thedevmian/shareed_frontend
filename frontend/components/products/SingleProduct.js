import { useQuery } from "@apollo/client";
import gql from "graphql-tag";

const GET_PRODUCT = gql`
  query GET_PRODUCT($id: ID!) {
    product(where: { id: $id }) {
      id
      name
      description
      price
      photo {
        id
        image {
          filename
          publicUrlTransformed
        }
        altText
      }
    }
  }
`;

const SingleProduct = ({ query }) => {
  const { loading, error, data } = useQuery(GET_PRODUCT, {
    variables: { id: query.id },
    });
  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :{error.message}</p>;
  
  return (
    <div>
      <h1>{data.product.name}</h1>
      <p>{data.product.description}</p>
      <p>{data.product.price}</p>
      <img src={data.product.photo.image} alt={data.product.photo.altText} />
    </div>
  );
};

export default SingleProduct;
