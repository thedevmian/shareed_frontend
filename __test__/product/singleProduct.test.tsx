import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import "@testing-library/jest-dom";
import { GetProductDocument } from "../../graphql/types-and-hooks";
import SingleProduct from "components/Products/SingleProduct";

const fakeProduct = {
  id: "abc123",
  name: "Product Name",
  description: "Product Description",
  price: 100,
  user: {
    id: "abc123",
    name: "User Name",
  },
  photo: [
    {
      id: "abc123",
      image: {
        publicUrlTransformed: "image.jpg",
        filename: "image.jpg",
      },
      altText: "Product Name",
    },
  ],
};

const mocks = [
  {
    request: {
      query: GetProductDocument,
      variables: {
        id: "1",
      },
    },
    result: {
      data: {
        product: fakeProduct,
      },
    },
  },
];
const errorMock = [
  {
    request: {
      query: GetProductDocument,
      variables: {
        id: "2",
      },
    },
    result: {
      data: {
        product: null as any,
      },
    },
    error: {
      name: "Error",
      message: "Item not found",
    },
  },
];

describe("Single Product Page", () => {
  it("renders with proper data", async () => {
    render(
      <MockedProvider mocks={mocks}>
        <SingleProduct id="1" />
      </MockedProvider>
    );
    expect(await screen.findByText(/Product Name/i)).toBeInTheDocument();
    expect(await screen.findByText(/Product Description/i)).toBeInTheDocument();
    expect(await screen.findByText("$1")).toBeInTheDocument();
  });

  it("render finding id", async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SingleProduct id="1" />
      </MockedProvider>
    );
    await screen.findByTestId("single-product");
    debug();
    expect(container).toMatchSnapshot();
  });
  it("Error out when no data", async () => {
    const { container, debug } = render(
      <MockedProvider mocks={errorMock}>
        <SingleProduct id="2" />
      </MockedProvider>
    );

    await screen.findByTestId("error");
    debug();
    expect(container).toHaveTextContent("Error :Item not found");
  });
});
