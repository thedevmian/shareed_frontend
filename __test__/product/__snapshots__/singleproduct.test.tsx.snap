// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Product Page render finding id 1`] = `
<div>
  <main
    class="sc-2f4b17dc-0 ifjHNx"
    data-testid="single-product"
  >
    <section
      class="sc-2f4b17dc-8 kSXgdl"
    >
      <div
        class="sc-2f4b17dc-6 cJzlzs"
      >
        <div
          data-rmiz-wrap="visible"
        >
          <div
            class="sc-945e8696-0 inaGdb"
          >
            <img
              alt="image.jpg"
              src="image.jpg"
            />
          </div>
          <button
            aria-label="Zoom image"
            data-rmiz-btn-open="true"
          />
        </div>
      </div>
      <div
        class="sc-2f4b17dc-5 cZrnWJ"
      >
        <div
          class="sc-2f4b17dc-7 bjDaUJ"
        >
          <h1
            class="sc-2f4b17dc-1 kcUOJh"
          >
            Product Name
          </h1>
          <div
            class="sc-22a4236-0 gAxNLm"
          >
            <button
              class="sc-22a4236-1 bMLVPZ btn btn-light m-0 text-left"
              type="button"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
        </div>
        <p
          class="sc-2f4b17dc-2 fjKDXZ"
        >
          Product Description
        </p>
        <div
          class="sc-2f4b17dc-7 bjDaUJ"
        >
          <p
            class="sc-2f4b17dc-3 kpjkMM"
          >
            $1
          </p>
          <span
            class="sc-2f4b17dc-4 fAzVKE"
          >
            (VAT included)
          </span>
        </div>
        <p
          class="sc-2f4b17dc-2 fjKDXZ"
        >
          Mid-washed light indigo denim jeans.
          <br />
          Straight fit Classic waist
          <br />
          Straight leg 18 cm bottom leg width (Size 30) 5 pockets Tobacco topstitches
          <br />
          Buttoned fly
          <br />
          Metal rivets and buttons in antic black color ami black leather patch on the back
          <br />
          Eco-friendly washing reducing its environmental impact
          <br />
          Main fabric:100% Cotton
        </p>
        <button
          class="sc-d8e035a9-0 hIoWhT"
        >
          Add to bag
        </button>
      </div>
    </section>
    <div
      class="sc-2f4b17dc-9 igxKcB"
    />
    <div>
      <h1
        class="sc-2f4b17dc-1 kcUOJh"
      >
        You might also like
      </h1>
    </div>
  </main>
</div>
`;
